### PR TITLE
[Qwen-Image] Drop unused vision tower from text encoder

### PR DESCRIPTION
**Pull request by lulugoodcoder**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose
Drops Qwen2.5-VL's unused vision tower from the Qwen-Image text encoder, freeing ~7 GB of GPU memory.

Qwen-Image's text encoder is `Qwen2_5_VLForConditionalGeneration`, which loads a vision tower (`model.visual`) that is never invoked during text-to-image generation. Deleting it immediately after `from_pretrained` is a pure memory optimization and does not affect generation quality.

### ⚠️  Upstream finding while testing — `--quantization fp8` produces dark images on A100 SXM4

  While testing this PR I also ran `--quantization fp8` and observed that **the existing FP8 path in main produces a dark / NaN image on NVIDIA A100 SXM4 (Ampere) for Qwen-Image**, independent of any change in this PR

Isolated by running `--quantization fp8` with `del visual` toggled both ways — both produce dark output.

## Test Plan 
Run BF16 baseline and FP8 inference end-to-end on Qwen-Image at 1024×1024, comparing memory and image quality.

  ```bash
  # BF16 baseline (this PR's only meaningful path on A100)
  python examples/offline_inference/text_to_image/text_to_image.py \
      --model Qwen/Qwen-Image --width 1024 --height 1024
 ```

 Visually inspect the saved qwen_image_output.png after each run.

## Test Result
 Hardware: NVIDIA A100 SXM4 80GB (Vast.ai), Qwen/Qwen-Image, 1024×1024, 50 inference steps, prompt "a cup of coffee
   on the table".

```  ┌───────────────────────────────────────────┬──────────────┬───────────────────────────────────────┐
  │                    Run                    │ Peak GPU mem │            Generated image            │
  ├───────────────────────────────────────────┼──────────────┼───────────────────────────────────────┤
  │ BF16 baseline (main, before this PR)      │ 53.7 GB      │ clear coffee cup ✓                    │
  ├───────────────────────────────────────────┼──────────────┼───────────────────────────────────────┤
  │ BF16 + del visual (this PR)               │ ~46.5 GB     │ clear coffee cup ✓                    │
  ├───────────────────────────────────────────┼──────────────┼───────────────────────────────────────┤
  │ FP8 + del visual (this PR)                │ 38.0 GB      │ dark / NaN ✗ (upstream — see Purpose) │
  ├───────────────────────────────────────────┼──────────────┼───────────────────────────────────────┤
  │ FP8 without del visual (this PR reverted) │ ~45 GB       │ dark / NaN ✗ (upstream — see Purpose) │
  └───────────────────────────────────────────┴──────────────┴───────────────────────────────────────┘
```
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [ ] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by hsliuustc0106** [comment]
does it affect qwen-image-edit?

**Comment by hsliuustc0106** [comment]
does it affect the accuracy?

**Comment by lulugoodcoder** [comment]
> does it affect the accuracy?

Good question. 
No, this PR does not affect Qwen-Image-Edit. Qwen-Image-Edit uses a separate pipeline file (pipeline_qwen_image_edit.py) which this PR does not modify

**Comment by lulugoodcoder** [comment]
> does it affect the accuracy?

  On accuracy — `del model.visual` is safe for T2I because the vision tower
  is unreachable from any code path in this pipeline. The full call trace:

  **1.** At `pipeline_qwen_image.py:437`, the text encoder is called with
  only text inputs:
  ```python
  encoder_hidden_states = self.text_encoder(
      input_ids=txt_tokens.input_ids,
      attention_mask=txt_tokens.attention_mask,
      output_hidden_states=True,
  )
```
  No pixel_values, no pixel_values_videos.

  2. This dispatches to
  Qwen2_5_VLForConditionalGeneration.forward(...) in transformers
  (modeling_qwen2_5_vl.py in https://github.com/huggingface/transformers/blob/main/src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py).
  Both pixel_values and pixel_values_videos default to None since we
  don't pass them.

  3. Inside that forward, self.visual(...) is reached only via:
  - get_image_features(pixel_values, ...) — guarded by if pixel_values is not None:
  - get_video_features(pixel_values_videos, ...) — guarded by if pixel_values_videos is 
  not None:

  With both inputs None, neither helper is called, so self.visual is
  never invoked.

  For contrast, the EDIT variant (pipeline_qwen_image_edit.py in this
  repo) DOES pass pixel_values
  and image_grid_thw. That pipeline needs the vision tower and is
  intentionally not modified by this PR.

  So del model.visual after from_pretrained is structurally equivalent
  to those parameters never being loaded

**Comment by lulugoodcoder** [comment]
@hsliuustc0106 I see several tests failing on this PR, but the
  `matches_diffusers` accuracy tests (which validate output of the exact
  T2I pipeline this PR modifies) passed in the same run. I'm not sure
  whether the other failures were caused by my change or are flaky / pre-
  existing — could you point me to a recent main build that ran the same
  tests so I can compare? Thanks!

**Comment by hsliuustc0106** [comment]
> @hsliuustc0106 I see several tests failing on this PR, but the `matches_diffusers` accuracy tests (which validate output of the exact T2I pipeline this PR modifies) passed in the same run. I'm not sure whether the other failures were caused by my change or are flaky / pre- existing — could you point me to a recent main build that ran the same tests so I can compare? Thanks!

the failures are known issues

**Comment by lulugoodcoder** [comment]
> > @hsliuustc0106 I see several tests failing on this PR, but the `matches_diffusers` accuracy tests (which validate output of the exact T2I pipeline this PR modifies) passed in the same run. I'm not sure whether the other failures were caused by my change or are flaky / pre- existing — could you point me to a recent main build that ran the same tests so I can compare? Thanks!
> 
> the failures are known issues

got it. thanks

**Comment by hsliuustc0106** [comment]
## PR Verification on NVIDIA L20X (143 GB)

Tested this PR end-to-end on real hardware using `Qwen/Qwen-Image`, 1024×1024, 50 inference steps.

### Results

| Run | Peak GPU Memory | Gen Time |
|---|---|---|
| main (baseline) | 59.3 GiB | 7.58s |
| PR (del visual) | **57.9 GiB** | 7.54s |
| **Delta** | **−1.2 GiB** | — |

Both runs produced correct output images with no quality difference.

### Notes

- Memory savings confirmed, though ~1.2 GiB on L20X vs the ~7 GiB reported on A100 — likely architecture-dependent allocation patterns for the vision tower weights.
- No functional regression — text-to-image generation works identically.
- The `torch_dtype=od_config.dtype` addition in `from_pretrained` is a nice touch to ensure consistent precision.

LGTM.

**Comment by lulugoodcoder** [comment]
> ## PR Verification on NVIDIA L20X (143 GB)
> Tested this PR end-to-end on real hardware using `Qwen/Qwen-Image`, 1024×1024, 50 inference steps.
> 
> ### Results
> Run	Peak GPU Memory	Gen Time
> main (baseline)	59.3 GiB	7.58s
> PR (del visual)	**57.9 GiB**	7.54s
> **Delta**	**−1.2 GiB**	—
> Both runs produced correct output images with no quality difference.
> 
> ### Notes
> * Memory savings confirmed, though ~1.2 GiB on L20X vs the ~7 GiB reported on A100 — likely architecture-dependent allocation patterns for the vision tower weights.
> * No functional regression — text-to-image generation works identically.
> * The `torch_dtype=od_config.dtype` addition in `from_pretrained` is a nice touch to ensure consistent precision.
> 
> LGTM.

Thanks for the detailed verification!

File: vllm_omni/diffusion/models/qwen_image/pipeline_qwen_image.py [hunk 1]
```diff
@@ -289,8 +289,15 @@ def __init__(
             model, subfolder="scheduler", local_files_only=local_files_only
         )
         self.text_encoder = Qwen2_5_VLForConditionalGeneration.from_pretrained(
-            model, subfolder="text_encoder", local_files_only=local_files_only
+            model,
+            subfolder="text_encoder",
+            local_files_only=local_files_only,
+            torch_dtype=od_config.dtype,
         ).to(self.device)
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/qwen_image/pipeline_qwen_image.py]: The full model (vision tower included) is moved to GPU here before the `del` on the next lines, so you still hit the ~7 GB spike on-device and only return the blocks to the caching allocator, not the OS. Deleting `model.visual` before `.to(self.device)` avoids the transient GPU spike entirely.

File: vllm_omni/diffusion/models/qwen_image/pipeline_qwen_image.py [hunk 2]
```diff
@@ -289,8 +289,15 @@ def __init__(
             model, subfolder="scheduler", local_files_only=local_files_only
         )
         self.text_encoder = Qwen2_5_VLForConditionalGeneration.from_pretrained(
-            model, subfolder="text_encoder", local_files_only=local_files_only
+            model,
+            subfolder="text_encoder",
+            local_files_only=local_files_only,
+            torch_dtype=od_config.dtype,
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/qwen_image/pipeline_qwen_image.py]: `torch_dtype=od_config.dtype` is a separate behavior change from dropping the vision tower (the encoder previously loaded in its checkpoint default dtype). Intentional to bundle it here?

Comment (lulugoodcoder) [reply vllm_omni/diffusion/models/qwen_image/pipeline_qwen_image.py]: yes, i want to ensure the consistent precision.

File: vllm_omni/diffusion/models/qwen_image/pipeline_qwen_image.py (hunk 1 above)
Comment (lulugoodcoder) [reply vllm_omni/diffusion/models/qwen_image/pipeline_qwen_image.py]: Thanks for the comment. I will make a new revision to address the above

**Comment by lulugoodcoder** [comment]
Heads up: I rebased onto latest `main` and squashed into a single clean commit, then force-pushed
1. **`del model.visual` before `.to(self.device)`** — moved the deletion
     to while the model is still on CPU, so the vision tower never reaches
     GPU memory (avoids the transient spike rather than just freeing blocks
     to the caching allocator).

  2. **`torch_dtype=od_config.dtype`** — kept intentionally

File: vllm_omni/diffusion/models/qwen_image/pipeline_qwen_image.py [hunk 3]
```diff
@@ -289,8 +289,17 @@ def __init__(
...
+        )
+        # Qwen2.5-VL ships a vision tower that text-to-image does not use.
+        # Drop it while the model is still on CPU, before moving to GPU, so
+        # the vision tower never consumes GPU memory.
+        if hasattr(self.text_encoder, "model") and hasattr(self.text_encoder.model, "visual"):
+            del self.text_encoder.model.visual
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/qwen_image/pipeline_qwen_image.py]: On older transformers `visual` sits directly on `Qwen2_5_VLForConditionalGeneration` (`self.text_encoder.visual`), not under `.model`. There the `hasattr` guard is False and this silently no-ops, so the vision tower stays in memory and the PR achieves nothing — with no signal. The codebase already guards both transformers layouts elsewhere (the "transformers v5" comments). hsliuustc0106's −1.2 GiB run was on the pre-force-push revision, so it doesn't cover this path. Handle both:

```suggestion
        visual_owner = None
        if hasattr(self.text_encoder, "model") and hasattr(self.text_encoder.model, "visual"):
            visual_owner = self.text_encoder.model
        elif hasattr(self.text_encoder, "visual"):
            visual_owner = self.text_encoder
        if visual_owner is not None:
            del visual_owner.visual
        else:
            logger.warning("Qwen-Image: vision tower not found on text encoder; skipping drop")
```

Comment (lulugoodcoder) [reply vllm_omni/diffusion/models/qwen_image/pipeline_qwen_image.py]: 1. Dropped `torch_dtype=od_config.dtype` — suspect that `test_qwen_image_matches_diffusers`
     fell to SSIM 0.967 (threshold 0.970) with it

  2. Applied your suggested fix handling both transformers layouts (`.model.visual`
     vs `.visual`) with the warning fallback so it can't silently no-op.

**Comment by lishunyang12** [comment]
LGTM now

**Review by lishunyang12 — APPROVED** [review]

**Comment by Semmer2** [comment]
LGTM, Qwen-Image generated identical images with and without this PR.
image

**Comment by princepride** [comment]
@hsliuustc0106 CI failed but seems not because this change.
